[XLA:GPU][sycl] Enable building GPU targets

akhilgoe opened the pull request:
The goal of this PR is to enable building all targets within `xla/stream_executor/` and `xla/service/gpu/` for sycl backend, without disrupting the code flow for the existing CUDA and ROCM backends. To achieve this, the PR:
1. Swaps (or inserts) some conditional guards with ones that help differentiate sycl with other gpu backends, and 
2. Adds stub implementations as placeholders, which will be completed in future PRs.

akhilgoe reviewed (commented):
Hi @beckerhe, just a gentle reminder to take a look at the PR when you get time. Thanks!

beckerhe commented:
> Hi @beckerhe, just a gentle reminder to take a look at the PR when you get time. Thanks!

Apologies. I had the review started, got distracted, and forgot about it. Looking now.

beckerhe commented on xla/backends/gpu/codegen/triton/BUILD:
```diff
@@ -174,7 +175,8 @@ cc_library(
     name = "fusion_emitter",
-    srcs = if_gpu_is_configured(
+    # Using if_cuda_or_rocm guard to prevent sycl target build / link errors.
+    srcs = if_cuda_or_rocm(
```
Unfortunately we can't use `if_cuda_or_rocm` here because internally it still means something else than `if_gpu_is_configured`.

Can you instead define a `if_cuda_or_rocm_is_configured` in xla/stream_executor/build_defs.bzl that does exactly the same in its implementation as `if_cuda_or_rocm`, then I can I wire this up correctly internally as well.

akhilgoe replied:
Thanks @beckerhe for the review!
There is a definition of `if_cuda_or_rocm` in `xla/stream_executor/build_defs.bzl` which re-uses the implementation from `local_config_rocm/rocm/build_defs.bzl`. Just double checking, you are asking me not to use it and to define `if_cuda_or_rocm_is_configured`, is that correct understanding?
https://github.com/openxla/xla/blob/b1ccbbc868434cc52707a750ab6c5b3302e93187/xla/stream_executor/build_defs.bzl#L28-L29

beckerhe replied:
Yes. In the same file and with the same implementation. It's just that we internally can redirect it to a different implementation.

beckerhe commented on xla/backends/gpu/collectives/BUILD:
```diff
@@ -227,8 +227,9 @@ cc_library(
-    srcs = ["nccl_errors.cc"],
-    hdrs = ["nccl_errors.h"],
+    # Using if_cuda_or_rocm guard to prevent sycl target build / link errors.
+    srcs = if_cuda_or_rocm(["nccl_errors.cc"]),
+    hdrs = if_cuda_or_rocm(["nccl_errors.h"]),
```
We're trying to actively get away from using these guards on attributes and rather make entire targets build for a certain platform.

Can you rather add a tag to this target - maybe `no-sycl` - and then filter out this tag through bazel?

(You might need to add some `if_cuda_or_rocm_is_configured` guards in the dependencies of targets that depend on this since the tag filter won't help for targets that are depended on.)

akhilgoe replied:
Done

beckerhe commented on xla/backends/gpu/collectives/BUILD:
```diff
@@ -251,8 +252,8 @@ cc_library(
     name = "nccl_collectives",
-    srcs = ["nccl_collectives.cc"],
-    hdrs = ["nccl_collectives.h"],
+    srcs = if_cuda_or_rocm(["nccl_collectives.cc"]),
+    hdrs = if_cuda_or_rocm(["nccl_collectives.h"]),
```
Same as above

akhilgoe replied:
Done

beckerhe commented on xla/backends/gpu/collectives/BUILD:
```diff
@@ -314,8 +315,8 @@ cc_library(
     name = "nccl_communicator",
-    srcs = ["nccl_communicator.cc"],
-    hdrs = ["nccl_communicator.h"],
+    srcs = if_cuda_or_rocm(["nccl_communicator.cc"]),
+    hdrs = if_cuda_or_rocm(["nccl_communicator.h"]),
```
Same as above

akhilgoe replied:
Done

beckerhe commented on xla/backends/gpu/collectives/BUILD:
```diff
@@ -411,7 +412,8 @@ cc_library(
 xla_test(
     name = "nccl_communicator_test",
-    srcs = ["nccl_communicator_test.cc"],
+    # Using if_cuda_or_rocm guard to prevent sycl target build / link errors.
+    srcs = if_cuda_or_rocm(["nccl_communicator_test.cc"]),
```
One more

akhilgoe replied:
Done

beckerhe commented on xla/service/gpu/BUILD:
```diff
@@ -2374,10 +2381,10 @@ cc_library(
         "@tsl//tsl/platform:ml_dtypes",
         "@tsl//tsl/platform:status",
         "@tsl//tsl/platform:statusor",
-    ] + if_gpu_is_configured(
+    ] + if_cuda_or_rocm(
```
if_cuda_or_rocm_is_configured as discussed before.

akhilgoe replied:
Done

beckerhe commented on xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1049,6 +1052,8 @@ absl::Status IrEmitterUnnested::EmitCholeskyThunk(const HloInstruction* instr) {
 
 #if GOOGLE_CUDA
   auto solver_creator = stream_executor::CudaSolverContext::Create;
+#elif TENSORFLOW_USE_SYCL
```
Where is this macro being set? Can you make it a local_define on the target?

Context: We're trying to remove all these ifdefs from the code. This one has not yet been refactored, so it's fine for now, but let's try not to introduce preprocessor defines with too big of a scope.

Also the `TENSORFLOW_` macro "namespacing" is outdated by now. Can you use `XLA_USE_SYCL` instead?

akhilgoe replied:
I believe it gets set while configuring the SYCL toolchain:

https://github.com/openxla/xla/blob/e3449ec355a344b30b49c2bfd49512801382ddd4/third_party/gpus/sycl_configure.bzl#L480-L483


Done, changed to a local define, `XLA_USE_SYCL`

beckerhe commented on xla/stream_executor/gpu/BUILD:
```diff
@@ -392,6 +392,24 @@ rocm_library(
+        "redzone_allocator_kernel.h",
+        "redzone_allocator_kernel_stub.cc",
+    ],
+    tags = ["manual"],
```
The manual tag is not needed here, since this should build just fine on any platform. If you remove it, then this target will be built on the CI and we make sure that it doesn't break.

akhilgoe replied:
Done, wasn't sure if you'd be interested in running SYCL on CI.

beckerhe reviewed (commented):
Thank you. In general looks good to me.

I've seen that this PR so far doesn't introduce any dependencies on SYCL headers.

I assume this is planned, right? It would be great if we introduced that in a way that we can at least build all SYCL targets on our CI - meaning it shouldn't require some specially set up system. Bazel can download dependencies/libraries as needed and that's how we also have the ROCm and the CUDA build set up.

akhilgoe commented on xla/service/gpu/build_defs.bzl:
```diff
@@ -59,16 +60,18 @@ def build_cub_sort_kernels(name, types, local_defines = [], **kwargs):
+def gpu_kernel_library(name, copts = [], srcs = [], local_defines = [], tags = [], **kwargs):
     cuda_library(
         name = name + "_cuda",
+        srcs = if_sycl_build_is_configured([], srcs),
```
Done

akhilgoe commented on xla/stream_executor/sycl/BUILD:
```diff
@@ -85,3 +86,31 @@ cc_library(
+    hdrs = if_sycl_is_configured(["sycl_solver_context.h"]),
+    local_defines = [
+        "TENSORFLOW_USE_SYCL=1",
+    ],
```
Yes, removed.

akhilgoe commented on xla/stream_executor/sycl/BUILD:
```diff
@@ -85,3 +86,31 @@ cc_library(
+cc_library(
+    name = "sycl_solver_context",
+    srcs = if_sycl_is_configured(["sycl_solver_context.cc"]),
+    hdrs = if_sycl_is_configured(["sycl_solver_context.h"]),
```
Done, added the `sycl-only` tag.

akhilgoe commented:
Thanks @beckerhe for the review! I have added a commit to address your comments. Please take a look whenever possible. 


> I've seen that this PR so far doesn't introduce any dependencies on SYCL headers.
> 
> I assume this is planned, right? It would be great if we introduced that in a way that we can at least build all SYCL targets on our CI - meaning it shouldn't require some specially set up system. Bazel can download dependencies/libraries as needed and that's how we also have the ROCm and the CUDA build set up.

Yes, we don't introduce any SYCL headers in this PR. The plan is to add them gradually. Glad to hear that you'd be interested in building SYCL targets on the CI! It would be easier to catch bugs that way.

Also, I defined `if_cuda_or_rocm_is_configured` as you suggested. Was wondering if `if_non_sycl_gpu_platform` would be a better name but that would assume that CUDA, ROCM and SYCL are the only 3 GPU platforms, what do you think?

beckerhe commented on xla/stream_executor/sycl/BUILD:
```diff
@@ -85,3 +86,28 @@ cc_library(
+    tags = [
+        "gpu",
+        "manual",
+        "sycl-only",
```
The CI wants all tags to be documented. Here is the error message:

```
ValueError: Tag(s) {'no-sycl', 'sycl-only'} are undocumented! Please document them in `build_tools/lint/tags.py`.
```

Can you take care of that?

beckerhe reviewed (commented):
In general looks good to me. I've just added two comments regarding linter nits.

akhilgoe commented on xla/backends/gpu/collectives/BUILD:
```diff
@@ -1,7 +1,7 @@
 load("@local_config_rocm//rocm:build_defs.bzl", "if_rocm_is_configured")
 load("//xla:xla.default.bzl", "xla_cc_test")
-load("//xla/stream_executor:build_defs.bzl", "if_gpu_is_configured")
+load("//xla/stream_executor:build_defs.bzl", "if_cuda_or_rocm_is_configured", "if_gpu_is_configured")
```
Done
